File: ontrack-extension-jenkins/src/test/java/net/nemerosa/ontrack/extension/jenkins/model/JenkinsJobPropertyTest.java
package net.nemerosa.ontrack.extension.jenkins.model;

import net.nemerosa.ontrack.extension.jenkins.JenkinsConfiguration;
import net.nemerosa.ontrack.extension.jenkins.JenkinsJobProperty;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class JenkinsJobPropertyTest {

    private JenkinsConfiguration configuration;
    private JenkinsJobProperty property;

    @Before
    public void before() {
        configuration = new JenkinsConfiguration(
                "MyConfig",
                "http://jenkins",
                "user",
                "secret"
        );
    }

    @Test
    public void url_simple() {
        property = new JenkinsJobProperty(
                configuration,
                "MyJob"
        );
        assertEquals("http://jenkins/job/MyJob", property.getUrl());
    }

    @Test
    public void url_folder_1_slash() {
        property = new JenkinsJobProperty(
                configuration,
                "prj/build"
        );
        assertEquals("http://jenkins/job/prj/job/build", property.getUrl());
    }

    @Test
    public void url_folder_2_slash() {
        property = new JenkinsJobProperty(
                configuration,
                "prj/branch/build"
        );
        assertEquals("http://jenkins/job/prj/job/branch/job/build", property.getUrl());
    }

    @Test
    public void url_folder_1_job() {
        property = new JenkinsJobProperty(
                configuration,
                "prj/job/build"
        );
        assertEquals("http://jenkins/job/prj/job/build", property.getUrl());
    }

    @Test
    public void url_folder_2_job() {
        property = new JenkinsJobProperty(
                configuration,
                "prj/job/branch/job/build"
        );
        assertEquals("http://jenkins/job/prj/job/branch/job/build", property.getUrl());
    }

}

File: ontrack-extension-jenkins/src/test/java/net/nemerosa/ontrack/extension/jenkins/JenkinsJobPropertyTest.kt
package net.nemerosa.ontrack.extension.jenkins

import org.junit.Test
import kotlin.test.assertEquals

class JenkinsJobPropertyTest {

    @Test
    fun getUrlForAnOrganizationBranch() {
        val property = JenkinsJobProperty(
            JenkinsConfiguration("Test", "https://host", "", ""),
            "organization/job/Repository/job/Branch"
        )
        assertEquals(
            "https://host/job/organization/job/Repository/job/Branch",
            property.url
        )
    }

    @Test
    fun getUrlForAnOrganizationBranchWithTrailingSlash() {
        val property = JenkinsJobProperty(
            JenkinsConfiguration("Test", "https://host", "", ""),
            "organization/job/Repository/job/Branch/"
        )
        assertEquals(
            "https://host/job/organization/job/Repository/job/Branch",
            property.url
        )
    }

    @Test
    fun getPathComponentsForSimpleJob() {
        val property = JenkinsJobProperty(
            JenkinsConfiguration("Test", "http://jenkins", "", ""),
            "test"
        )
        assertEquals(listOf("test"), property.pathComponents)
    }

    @Test
    fun getPathComponentsForFolderJob() {
        val property = JenkinsJobProperty(
            JenkinsConfiguration("Test", "http://jenkins", "", ""),
            "test/test-master/test-master-build"
        )
        assertEquals(listOf("test", "test-master", "test-master-build"), property.pathComponents)
    }

    @Test
    fun getPathComponentsForCompleteFolderJob() {
        val property = JenkinsJobProperty(
            JenkinsConfiguration("Test", "http://jenkins", "", ""),
            "test/job/test-master/job/test-master-build"
        )
        assertEquals(listOf("test", "test-master", "test-master-build"), property.pathComponents)
    }
}